frontend/src/services/transactions/fill-manager.ts

```diff
@@ -1,46 +1,48 @@
-import type { JsonRpcProvider } from '@ethersproject/providers';
+import { JsonRpcProvider } from '@ethersproject/providers';
 import { BigNumber, Contract } from 'ethers';
 
 import FillManager from '@/assets/FillManager.json';
+import type { Cancelable } from '@/types/async';
 import type { UInt256 } from '@/types/uint-256';
 
-export async function waitForFulfillment(
-  targetNetworkProvider: JsonRpcProvider,
-  requestIdentifier: UInt256,
+export async function checkForPastFulfillmentEvent(
+  rpcUrl: string,
   fillManagerAddress: string,
-): Promise<void> {
-  const fillManagerContract = new Contract(
-    fillManagerAddress,
-    FillManager.abi,
-    targetNetworkProvider,
-  );
-
-  const currentBlockNumber = await targetNetworkProvider.getBlockNumber();
-  const eventFilter = fillManagerContract.filters.RequestFilled(
-    BigNumber.from(requestIdentifier.asString),
-  );
-  const events = await fillManagerContract.queryFilter(eventFilter, currentBlockNumber - 500);
-  if (events.length > 0) return;
-
-  const eventListeningTimeout = 1000 * 30;
+  requestIdentifier: UInt256,
+): Promise<boolean> {
+  const provider = new JsonRpcProvider(rpcUrl);
+  const contract = new Contract(fillManagerAddress, FillManager.abi, provider);
+  const currentBlockNumber = await provider.getBlockNumber();
+  const eventFilter = contract.filters.RequestFilled(BigNumber.from(requestIdentifier.asString));
+  const events = await contract.queryFilter(eventFilter, currentBlockNumber - 500);
+  return events.length > 0;
+}
 
-  const fulfillmentPromise: Promise<void> = new Promise((resolve) => {
-    fillManagerContract.on(eventFilter, () => {
+export function waitForFulfillment(
+  rpcUrl: string,
+  fillManagerAddress: string,
+  requestIdentifier: UInt256,
+): Cancelable<void> {
+  const provider = new JsonRpcProvider(rpcUrl);
+  const contract = new Contract(fillManagerAddress, FillManager.abi, provider);
```

Comment:
Okay, I can do that. This will be refactored soon anyway, but I like such recommendations and thereby follow it. Always good to keep the code base clean, even if you plan to remove stuff soonish.

Reply:
It was less effective than I though due to the mixmax with signers and sometimes functions need the contract but also the provider. Anyway, I did it.

Reply:
✔️